File: common/src/test/java/DahRealPanda/plugins/colonyweb/auth/AuthServiceTest.java
package DahRealPanda.plugins.colonyweb.auth;

import DahRealPanda.plugins.colonyweb.Config;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * The sign-in flow: pairing codes issued in game, exchanged for a session, and the access
 * control that decides which colonies a session may read.
 *
 * <p>Because a code is the only credential in the system, the unhappy paths matter more than
 * the happy one — a code that could be redeemed twice, survive expiry, or be guessed
 * case-by-case would each be a way into somebody else's colony.</p>
 */
class AuthServiceTest {

    @TempDir
    Path dataDir;

    private static final UUID ALICE = UUID.fromString("11111111-2222-3333-4444-555555555555");
    private static final UUID BOB = UUID.fromString("66666666-7777-8888-9999-000000000000");

    @BeforeEach
    void configureDefaults() {
        // Config is plain static state filled in by the loader at runtime; tests set it directly.
        Config.authEnabled = true;
        Config.sessionDays = 30;
        Config.loginCodeMinutes = 10;
    }

    private AuthService service() {
        return new AuthService(dataDir);
    }

    // ------------------------------------------------------------------
    // Happy path
    // ------------------------------------------------------------------

    @Nested
    @DisplayName("signing in")
    class SigningIn {

        @Test
        @DisplayName("a freshly issued code exchanges for a session that resolves to the player")
        void codeRedeemsToSession() {
            AuthService auth = service();
            String code = auth.issueCode(ALICE, "Alice", List.of(1, 2), false);

            Optional<String> token = auth.redeemCode(code);

            assertTrue(token.isPresent(), "a fresh code should redeem");
            WebUser user = auth.userForToken(token.get()).orElseThrow();
            assertEquals("Alice", user.name);
            assertEquals(Set.of(1, 2), user.accessibleColonies());
        }

        @Test
        @DisplayName("codes are formatted XXXX-XXXX and avoid look-alike characters")
        void codeShape() {
            String code = service().issueCode(ALICE, "Alice", List.of(), false);

            assertTrue(code.matches("[23456789BCDFGHJKLMNPQRSTVWXZ]{4}-[23456789BCDFGHJKLMNPQRSTVWXZ]{4}"),
                    "unexpected code shape: " + code);
        }

        @Test
        @DisplayName("two players get separate sessions")
        void sessionsAreNotShared() {
            AuthService auth = service();
            String aliceToken = auth.redeemCode(auth.issueCode(ALICE, "Alice", List.of(1), false)).orElseThrow();
            String bobToken = auth.redeemCode(auth.issueCode(BOB, "Bob", List.of(2), false)).orElseThrow();

            assertNotEquals(aliceToken, bobToken);
            assertEquals("Alice", auth.userForToken(aliceToken).orElseThrow().name);
            assertEquals("Bob", auth.userForToken(bobToken).orElseThrow().name);
        }

        @Test
        @DisplayName("accounts and sessions survive a restart")
        void persistsAcrossRestart() {
            AuthService before = service();
            String token = before.redeemCode(before.issueCode(ALICE, "Alice", List.of(7), false))
                    .orElseThrow();

            // A second instance over the same data directory is what a server restart looks like.
            AuthService after = service();

            assertEquals("Alice", after.userForToken(token).orElseThrow().name);
            assertEquals(Set.of(7), after.user(ALICE).orElseThrow().accessibleColonies());
        }

        @Test
        @DisplayName("a pending code does not survive a restart")
        void pendingCodesAreNotPersisted() {
            String code = service().issueCode(ALICE, "Alice", List.of(1), false);

            // Deliberate: codes live in memory only, so a restart cancels half-finished logins
            // rather than leaving a credential valid across it.
            assertTrue(service().redeemCode(code).isEmpty());
        }

        @Test
        @DisplayName("the account file is written to the data directory")
        void writesAuthJson() {
            service().issueCode(ALICE, "Alice", List.of(1), false);

            assertTrue(Files.isRegularFile(dataDir.resolve("auth.json")));
        }
    }

    // ------------------------------------------------------------------
    // Codes that must not work
    // ------------------------------------------------------------------

    @Nested
    @DisplayName("rejecting bad codes")
    class RejectingBadCodes {

        @Test
        @DisplayName("a code that was never issued is refused")
        void unknownCode() {
            assertTrue(service().redeemCode("ZZZZ-ZZZZ").isEmpty());
        }

        @Test
        @DisplayName("a code cannot be redeemed twice")
        void singleUse() {
            AuthService auth = service();
            String code = auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertTrue(auth.redeemCode(code).isPresent());
            assertTrue(auth.redeemCode(code).isEmpty(), "the second redemption should fail");
        }

        @Test
        @DisplayName("an expired code is refused")
        void expiredCode() {
            AuthService auth = service();
            // Zero minutes means the code expires the instant it is issued.
            Config.loginCodeMinutes = 0;
            String code = auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertTrue(auth.redeemCode(code).isEmpty());
        }

        @Test
        @DisplayName("re-syncing invalidates the previously issued code")
        void reissueInvalidatesOldCode() {
            AuthService auth = service();
            String first = auth.issueCode(ALICE, "Alice", List.of(1), false);
            String second = auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertTrue(auth.redeemCode(first).isEmpty(), "the superseded code should be dead");
            assertTrue(auth.redeemCode(second).isPresent());
        }

        @ParameterizedTest
        @DisplayName("null, blank and malformed codes are refused rather than throwing")
        @ValueSource(strings = {"", "   ", "----", "not a code", "AAAA-AAAA-AAAA"})
        void malformedCode(String code) {
            assertTrue(service().redeemCode(code).isEmpty());
        }

        @Test
        @DisplayName("a null code is refused")
        void nullCode() {
            assertTrue(service().redeemCode(null).isEmpty());
        }
    }

    @Nested
    @DisplayName("code normalisation")
    class CodeNormalisation {

        @Test
        @DisplayName("case does not matter")
        void caseInsensitive() {
            AuthService auth = service();
            String code = auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertTrue(auth.redeemCode(code.toLowerCase(java.util.Locale.ROOT)).isPresent());
        }

        @Test
        @DisplayName("the dash is optional")
        void dashOptional() {
            AuthService auth = service();
            String code = auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertTrue(auth.redeemCode(code.replace("-", "")).isPresent());
        }

        @Test
        @DisplayName("stray spaces are ignored, so a pasted code still works")
        void spacesIgnored() {
            AuthService auth = service();
            String code = auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertTrue(auth.redeemCode(" " + code + " ").isPresent());
        }
    }

    // ------------------------------------------------------------------
    // Sessions
    // ------------------------------------------------------------------

    @Nested
    @DisplayName("sessions")
    class Sessions {

        @Test
        @DisplayName("an unknown token resolves to nobody")
        void unknownToken() {
            assertTrue(service().userForToken("deadbeef").isEmpty());
        }

        @ParameterizedTest
        @DisplayName("null and blank tokens resolve to nobody")
        @ValueSource(strings = {"", " ", "\t"})
        void blankToken(String token) {
            assertTrue(service().userForToken(token).isEmpty());
        }

        @Test
        @DisplayName("a null token resolves to nobody")
        void nullToken() {
            assertTrue(service().userForToken(null).isEmpty());
        }

        @Test
        @DisplayName("an expired session no longer resolves")
        void expiredSession() {
            AuthService auth = service();
            Config.sessionDays = 0;
            String token = auth.redeemCode(auth.issueCode(ALICE, "Alice", List.of(1), false)).orElseThrow();

            assertTrue(auth.userForToken(token).isEmpty());
        }

        @Test
        @DisplayName("signing out drops only that browser's session")
        void revokeOneToken() {
            AuthService auth = service();
            String first = auth.redeemCode(auth.issueCode(ALICE, "Alice", List.of(1), false)).orElseThrow();
            String second = auth.redeemCode(auth.issueCode(ALICE, "Alice", List.of(1), false)).orElseThrow();

            auth.revokeToken(first);

            assertTrue(auth.userForToken(first).isEmpty(), "the signed-out browser should be gone");
            assertTrue(auth.userForToken(second).isPresent(), "the other browser should still work");
        }

        @Test
        @DisplayName("revoking an unknown token changes nothing")
        void revokeUnknownToken() {
            AuthService auth = service();
            String token = auth.redeemCode(auth.issueCode(ALICE, "Alice", List.of(1), false)).orElseThrow();

            auth.revokeToken("not-a-real-token");
            auth.revokeToken(null);

            assertTrue(auth.userForToken(token).isPresent());
        }

        @Test
        @DisplayName("an operator can sign a player out everywhere")
        void revokeAll() {
            AuthService auth = service();
            String first = auth.redeemCode(auth.issueCode(ALICE, "Alice", List.of(1), false)).orElseThrow();
            String second = auth.redeemCode(auth.issueCode(ALICE, "Alice", List.of(1), false)).orElseThrow();

            assertEquals(2, auth.revokeAllSessions(ALICE));

            assertTrue(auth.userForToken(first).isEmpty());
            assertTrue(auth.userForToken(second).isEmpty());
        }

        @Test
        @DisplayName("signing out a player with no account reports nothing revoked")
        void revokeAllForUnknownPlayer() {
            assertEquals(0, service().revokeAllSessions(BOB));
        }

        @Test
        @DisplayName("purging drops expired sessions and leaves live ones alone")
        void purgeExpired() {
            AuthService auth = service();
            Config.sessionDays = 0;
            auth.redeemCode(auth.issueCode(ALICE, "Alice", List.of(1), false)).orElseThrow();
            Config.sessionDays = 30;
            String live = auth.redeemCode(auth.issueCode(BOB, "Bob", List.of(2), false)).orElseThrow();

            assertEquals(2, auth.sessionCount());
            auth.purgeExpiredSessions();

            assertEquals(1, auth.sessionCount());
            assertTrue(auth.userForToken(live).isPresent());
        }
    }

    // ------------------------------------------------------------------
    // Pending code bookkeeping
    // ------------------------------------------------------------------

    @Nested
    @DisplayName("pending code count")
    class PendingCodes {

        @Test
        @DisplayName("counts codes waiting to be typed in")
        void countsOutstanding() {
            AuthService auth = service();
            auth.issueCode(ALICE, "Alice", List.of(1), false);
            auth.issueCode(BOB, "Bob", List.of(2), false);

            assertEquals(2, auth.pendingCodeCount());
        }

        @Test
        @DisplayName("a redeemed code stops counting")
        void redeemedCodeIsGone() {
            AuthService auth = service();
            String code = auth.issueCode(ALICE, "Alice", List.of(1), false);
            auth.redeemCode(code);

            assertEquals(0, auth.pendingCodeCount());
        }

        @Test
        @DisplayName("expired codes are not counted")
        void expiredNotCounted() {
            AuthService auth = service();
            Config.loginCodeMinutes = 0;
            auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertEquals(0, auth.pendingCodeCount());
        }

        @Test
        @DisplayName("signing a player out everywhere also cancels their pending code")
        void revokeAllCancelsCode() {
            AuthService auth = service();
            String code = auth.issueCode(ALICE, "Alice", List.of(1), false);

            auth.revokeAllSessions(ALICE);

            assertEquals(0, auth.pendingCodeCount());
            assertTrue(auth.redeemCode(code).isEmpty());
        }
    }

    // ------------------------------------------------------------------
    // Access control
    // ------------------------------------------------------------------

    @Nested
    @DisplayName("colony access")
    class ColonyAccess {

        @Test
        @DisplayName("a member sees their own colony and no others")
        void memberColoniesOnly() {
            AuthService auth = service();
            auth.issueCode(ALICE, "Alice", List.of(1), false);
            WebUser alice = auth.user(ALICE).orElseThrow();

            assertTrue(auth.canAccess(alice, 1));
            assertFalse(auth.canAccess(alice, 2));
        }

        @Test
        @DisplayName("an operator sees every colony")
        void adminSeesEverything() {
            AuthService auth = service();
            auth.issueCode(ALICE, "Alice", List.of(), true);
            WebUser alice = auth.user(ALICE).orElseThrow();

            assertTrue(auth.canAccess(alice, 999));
        }

        @Test
        @DisplayName("nobody is refused while authentication is switched off")
        void authDisabledOpensEverything() {
            Config.authEnabled = false;

            assertTrue(service().canAccess(null, 1), "a disabled dashboard is deliberately open");
        }

        @Test
        @DisplayName("an anonymous caller is refused while authentication is on")
        void nullUserRefused() {
            assertFalse(service().canAccess(null, 1));
        }

        @Test
        @DisplayName("an explicit grant survives a re-sync that drops the membership")
        void grantSurvivesResync() {
            AuthService auth = service();
            auth.issueCode(ALICE, "Alice", List.of(1), false);
            assertTrue(auth.grant(ALICE, "Alice", 5));

            // A later sync replaces the mirrored colonies wholesale.
            auth.issueCode(ALICE, "Alice", List.of(1), false);
            WebUser alice = auth.user(ALICE).orElseThrow();

            assertTrue(auth.canAccess(alice, 5), "the operator's grant must not be lost");
            assertEquals(Set.of(1, 5), alice.accessibleColonies());
        }

        @Test
        @DisplayName("granting the same colony twice reports no change the second time")
        void grantIsIdempotent() {
            AuthService auth = service();

            assertTrue(auth.grant(ALICE, "Alice", 5));
            assertFalse(auth.grant(ALICE, "Alice", 5));
        }

        @Test
        @DisplayName("granting creates an account for a player who has never synced")
        void grantCreatesAccount() {
            AuthService auth = service();

            assertTrue(auth.grant(BOB, "Bob", 3));

            assertEquals(Set.of(3), auth.user(BOB).orElseThrow().accessibleColonies());
        }

        @Test
        @DisplayName("revoking a grant leaves in-game membership intact")
        void revokeGrantKeepsMembership() {
            AuthService auth = service();
            auth.issueCode(ALICE, "Alice", List.of(1), false);
            auth.grant(ALICE, "Alice", 5);

            assertTrue(auth.revokeGrant(ALICE, 5));

            WebUser alice = auth.user(ALICE).orElseThrow();
            assertFalse(auth.canAccess(alice, 5));
            assertTrue(auth.canAccess(alice, 1), "membership is not a grant and must survive");
        }

        @Test
        @DisplayName("revoking a grant that was never made reports no change")
        void revokeMissingGrant() {
            AuthService auth = service();
            auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertFalse(auth.revokeGrant(ALICE, 5));
            assertFalse(auth.revokeGrant(BOB, 5), "an unknown player has nothing to revoke");
        }
    }

    @Nested
    @DisplayName("account lookup")
    class AccountLookup {

        @Test
        @DisplayName("a player who never synced has no account")
        void unknownPlayer() {
            assertTrue(service().user(BOB).isEmpty());
        }

        @Test
        @DisplayName("lookup ignores UUID case")
        void uuidCaseInsensitive() {
            AuthService auth = service();
            auth.issueCode(ALICE, "Alice", List.of(1), false);

            assertTrue(auth.user(UUID.fromString(ALICE.toString().toUpperCase(java.util.Locale.ROOT))).isPresent());
        }

        @Test
        @DisplayName("every known account is listed")
        void listsAllUsers() {
            AuthService auth = service();
            auth.issueCode(ALICE, "Alice", List.of(1), false);
            auth.issueCode(BOB, "Bob", List.of(2), false);

            assertEquals(2, auth.allUsers().size());
        }

        @Test
        @DisplayName("a rename is picked up on the next sync")
        void nameIsRefreshed() {
            AuthService auth = service();
            auth.issueCode(ALICE, "OldName", List.of(1), false);
            auth.issueCode(ALICE, "NewName", List.of(1), false);

            assertEquals("NewName", auth.user(ALICE).orElseThrow().name);
        }
    }
}

File: common/src/test/kotlin/DahRealPanda/plugins/colonyweb/auth/AuthServiceTest.kt
package DahRealPanda.plugins.colonyweb.auth

import DahRealPanda.plugins.colonyweb.Config
import DahRealPanda.plugins.colonyweb.support.ConfigReset
import DahRealPanda.plugins.colonyweb.support.withTempDir
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldMatch
import io.kotest.matchers.string.shouldNotContain
import java.nio.file.Files
import java.nio.file.Path
import java.util.UUID

/**
 * [AuthService] is the only thing standing between a colony's data and anyone who can reach the
 * port, so both halves matter: the codes and sessions it accepts, and — more importantly — the
 * ones it refuses.
 */
class AuthServiceTest : DescribeSpec({

    val ann = UUID.fromString("00000000-0000-0000-0000-0000000000a1")
    val bob = UUID.fromString("00000000-0000-0000-0000-0000000000b2")

    beforeTest { ConfigReset.applyDefaults() }

    fun authIn(dir: Path) = AuthService(dir)

    describe("issuing a pairing code") {
        it("produces a code in the two-groups-of-four shape the dashboard asks for") {
            withTempDir { dir ->
                authIn(dir).issueCode(ann, "Ann", listOf(1), false) shouldMatch Regex("[A-Z0-9]{4}-[A-Z0-9]{4}")
            }
        }

        // The player reads this code off chat and types it into a browser, so anything that can be
        // confused for something else — vowels that spell words, 0/O, 1/I — is left out.
        it("uses no vowels and no look-alike characters") {
            withTempDir { dir ->
                val auth = authIn(dir)
                repeat(200) {
                    val code = auth.issueCode(ann, "Ann", listOf(1), false).replace("-", "")
                    code.forEach { character ->
                        ("23456789BCDFGHJKLMNPQRSTVWXZ".contains(character)) shouldBe true
                    }
                }
            }
        }

        it("issues a different code each time") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val codes = (1..50).map { auth.issueCode(ann, "Ann", listOf(1), false) }
                codes.toSet() shouldHaveSize codes.size
            }
        }

        it("invalidates the player's previous code, so only the newest one works") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val first = auth.issueCode(ann, "Ann", listOf(1), false)
                val second = auth.issueCode(ann, "Ann", listOf(1), false)

                auth.redeemCode(first) shouldBe null
                auth.redeemCode(second).shouldNotBeNull()
            }
        }

        it("does not invalidate another player's pending code") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val annsCode = auth.issueCode(ann, "Ann", listOf(1), false)
                auth.issueCode(bob, "Bob", listOf(2), false)

                auth.redeemCode(annsCode).shouldNotBeNull()
            }
        }

        it("records the colonies, admin flag and current name on the account") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1, 2), true)

                val user = auth.user(ann).shouldNotBeNull()
                user.name shouldBe "Ann"
                user.admin shouldBe true
                user.colonies shouldContainExactlyInAnyOrder listOf(1, 2)
                (user.syncedAt > 0) shouldBe true
            }
        }

        it("replaces the colony list on re-sync, so leaving a colony revokes access to it") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1, 2), false)
                auth.issueCode(ann, "Ann", listOf(2), false)

                auth.user(ann)!!.colonies shouldContainExactlyInAnyOrder listOf(2)
            }
        }

        it("picks up a rename on the next sync") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1), false)
                auth.issueCode(ann, "Annabel", listOf(1), false)

                auth.user(ann)!!.name shouldBe "Annabel"
            }
        }

        it("treats the same player id in a different case as the same account") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1), false)

                auth.user(UUID.fromString(ann.toString().uppercase())).shouldNotBeNull()
                auth.allUsers() shouldHaveSize 1
            }
        }
    }

    describe("redeeming a pairing code") {
        it("returns a session token for a valid code") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val code = auth.issueCode(ann, "Ann", listOf(1), false)

                val token = auth.redeemCode(code).shouldNotBeNull()
                auth.userForToken(token)!!.uuid shouldBe ann.toString()
            }
        }

        it("accepts the code however the player typed it") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val code = auth.issueCode(ann, "Ann", listOf(1), false)

                auth.redeemCode(code.replace("-", "").lowercase()).shouldNotBeNull()
            }
        }

        it("ignores stray spaces around and inside the code") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val code = auth.issueCode(ann, "Ann", listOf(1), false)

                auth.redeemCode("  ${code.replace("-", " - ")}  ").shouldNotBeNull()
            }
        }

        it("burns the code, so a shoulder-surfer cannot reuse it") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val code = auth.issueCode(ann, "Ann", listOf(1), false)

                auth.redeemCode(code).shouldNotBeNull()
                auth.redeemCode(code) shouldBe null
            }
        }

        it("rejects a code that was never issued") {
            withTempDir { dir ->
                authIn(dir).redeemCode("ZZZZ-ZZZZ") shouldBe null
            }
        }

        it("rejects an empty and a blank code") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1), false)

                auth.redeemCode("") shouldBe null
                auth.redeemCode("   ") shouldBe null
            }
        }

        it("rejects a code that has expired") {
            withTempDir { dir ->
                val auth = authIn(dir)
                Config.loginCodeMinutes = 0
                val code = auth.issueCode(ann, "Ann", listOf(1), false)

                auth.redeemCode(code) shouldBe null
            }
        }

        it("stops counting a code as pending once it has expired") {
            withTempDir { dir ->
                val auth = authIn(dir)
                Config.loginCodeMinutes = 0
                auth.issueCode(ann, "Ann", listOf(1), false)

                auth.pendingCodeCount() shouldBe 0
            }
        }
    }

    describe("sessions") {
        it("resolves a live token to its owner") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val token = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!

                auth.userForToken(token)!!.name shouldBe "Ann"
                auth.sessionCount() shouldBe 1
            }
        }

        it("rejects a token nobody was ever given") {
            withTempDir { dir ->
                authIn(dir).userForToken("deadbeef") shouldBe null
            }
        }

        it("rejects a null and a blank token") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.userForToken(null) shouldBe null
                auth.userForToken("") shouldBe null
                auth.userForToken("   ") shouldBe null
            }
        }

        it("rejects a token whose session has run out") {
            withTempDir { dir ->
                val auth = authIn(dir)
                Config.sessionDays = 0
                val token = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!

                auth.userForToken(token) shouldBe null
            }
        }

        it("does not hand one player's token to another") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val annsToken = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!
                auth.redeemCode(auth.issueCode(bob, "Bob", listOf(2), false))!!

                auth.userForToken(annsToken)!!.name shouldBe "Ann"
            }
        }

        it("logging in twice leaves both browsers signed in") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val first = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!
                val second = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!

                first shouldNotBe second
                auth.userForToken(first).shouldNotBeNull()
                auth.userForToken(second).shouldNotBeNull()
                auth.sessionCount() shouldBe 2
            }
        }

        it("revoking a token signs out only that browser") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val first = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!
                val second = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!

                auth.revokeToken(first)

                auth.userForToken(first) shouldBe null
                auth.userForToken(second).shouldNotBeNull()
            }
        }

        it("revoking an unknown, null or blank token is a no-op") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val token = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!

                auth.revokeToken("not-a-token")
                auth.revokeToken(null)
                auth.revokeToken("")

                auth.userForToken(token).shouldNotBeNull()
            }
        }

        it("revoking everything reports how many sessions were dropped") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))
                val token = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!

                auth.revokeAllSessions(ann) shouldBe 2
                auth.userForToken(token) shouldBe null
            }
        }

        it("revoking everything for a player with no account reports nothing") {
            withTempDir { dir ->
                authIn(dir).revokeAllSessions(ann) shouldBe 0
            }
        }

        it("revoking everything also cancels a code that was never redeemed") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val code = auth.issueCode(ann, "Ann", listOf(1), false)

                auth.revokeAllSessions(ann)

                auth.redeemCode(code) shouldBe null
            }
        }

        it("purging drops expired sessions and leaves live ones alone") {
            withTempDir { dir ->
                val auth = authIn(dir)
                Config.sessionDays = 0
                auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))
                Config.sessionDays = 30
                val live = auth.redeemCode(auth.issueCode(bob, "Bob", listOf(2), false))!!

                auth.purgeExpiredSessions()

                auth.sessionCount() shouldBe 1
                auth.userForToken(live).shouldNotBeNull()
            }
        }
    }

    describe("what gets written to disk") {
        it("never stores the session token itself") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val token = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!

                val stored = Files.readString(dir.resolve("auth.json"))
                stored shouldNotContain token
                stored.contains("tokenHash") shouldBe true
            }
        }

        it("keeps accounts and sessions across a server restart") {
            withTempDir { dir ->
                val auth = authIn(dir)
                val token = auth.redeemCode(auth.issueCode(ann, "Ann", listOf(1), false))!!

                val reopened = authIn(dir)
                reopened.user(ann).shouldNotBeNull()
                reopened.userForToken(token).shouldNotBeNull()
                reopened.canAccess(reopened.user(ann), 1) shouldBe true
            }
        }

        // Pending codes are deliberately in-memory only: a code is valid for minutes, and a code
        // that outlived the restart it was issued before would be a credential nobody remembers.
        it("forgets pending codes across a restart") {
            withTempDir { dir ->
                val code = authIn(dir).issueCode(ann, "Ann", listOf(1), false)

                authIn(dir).redeemCode(code) shouldBe null
            }
        }
    }

    describe("access control") {
        it("lets a member see their own colony") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1), false)

                auth.canAccess(auth.user(ann), 1) shouldBe true
            }
        }

        it("refuses a colony the player does not belong to") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1), false)

                auth.canAccess(auth.user(ann), 2) shouldBe false
            }
        }

        it("refuses an anonymous visitor while auth is on") {
            withTempDir { dir ->
                authIn(dir).canAccess(null, 1) shouldBe false
            }
        }

        it("lets an operator see every colony") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", emptyList(), true)

                auth.canAccess(auth.user(ann), 99) shouldBe true
            }
        }

        // With auth off the dashboard is deliberately wide open, and the guard has to say so even
        // for a caller that never signed in.
        it("lets anyone in when auth is switched off entirely") {
            withTempDir { dir ->
                val auth = authIn(dir)
                Config.authEnabled = false

                auth.enabled() shouldBe false
                auth.canAccess(null, 1) shouldBe true
            }
        }

        it("honours a colony an operator granted by hand") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1), false)

                auth.grant(ann, "Ann", 7) shouldBe true
                auth.canAccess(auth.user(ann), 7) shouldBe true
            }
        }

        it("reports a repeated grant as a no-op") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.grant(ann, "Ann", 7) shouldBe true
                auth.grant(ann, "Ann", 7) shouldBe false
            }
        }

        it("can grant to a player who has never signed in") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.grant(ann, "Ann", 7) shouldBe true

                auth.user(ann)!!.name shouldBe "Ann"
            }
        }

        it("takes a granted colony away again") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.grant(ann, "Ann", 7)

                auth.revokeGrant(ann, 7) shouldBe true
                auth.canAccess(auth.user(ann), 7) shouldBe false
            }
        }

        it("reports revoking a grant that was never made") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.grant(ann, "Ann", 7)

                auth.revokeGrant(ann, 8) shouldBe false
                auth.revokeGrant(bob, 7) shouldBe false
            }
        }

        // A grant is the operator saying "yes, this one too". A sync mirrors MineColonies and
        // replaces the mirrored list wholesale, so it must not take the grant with it.
        it("keeps a grant across a re-sync that drops the player's colonies") {
            withTempDir { dir ->
                val auth = authIn(dir)
                auth.issueCode(ann, "Ann", listOf(1), false)
                auth.grant(ann, "Ann", 7)

                auth.issueCode(ann, "Ann", emptyList(), false)

                auth.canAccess(auth.user(ann), 7) shouldBe true
                auth.canAccess(auth.user(ann), 1) shouldBe false
            }
        }
    }
})
